masonry/src/passes/update.rs

```diff
@@ -0,0 +1,220 @@
+
+    let Some(parent_id) = parent_id else {
+        // We've reached the root
+        let child_state_mut = arena.widget_states.find_mut(widget_id.to_raw()).unwrap();
+        root_state.merge_up(child_state_mut.item);
+        return;
+    };
+
+    let mut parent_state_mut = arena.widget_states.find_mut(parent_id.to_raw()).unwrap();
+    let child_state_mut = parent_state_mut
+        .children
+        .get_child_mut(widget_id.to_raw())
+        .unwrap();
+
+    parent_state_mut.item.merge_up(child_state_mut.item);
+}
+
+fn get_widget_mut(arena: &mut WidgetArena, id: WidgetId) -> (&mut dyn Widget, PassCtx<'_>) {
+    let state_mut = arena
+        .widget_states
+        .find_mut(id.to_raw())
+        .expect("widget state not found in arena");
+    let widget_mut = arena
+        .widgets
+        .find_mut(id.to_raw())
+        .expect("widget not found in arena");
+
+    // Box<dyn Widget> -> &dyn Widget
+    // Without this step, the type of `WidgetRef::widget` would be
+    // `&Box<dyn Widget> as &dyn Widget`, which would be an additional layer
+    // of indirection.
+    let widget = widget_mut.item;
+    let widget: &mut dyn Widget = &mut **widget;
+
+    (
+        widget,
+        PassCtx {
+            widget_state: state_mut,
+            widget_children: widget_mut.children,
+        },
+    )
+}
+
+// TODO - Replace LifecycleCtx with UpdateCtx
+pub(crate) fn run_targeted_update_pass(
+    root: &mut RenderRoot,
+    root_state: &mut WidgetState,
+    target: Option<WidgetId>,
+    pass_fn: impl FnMut(&mut dyn Widget, &mut LifeCycleCtx),
+) {
+    let mut pass_fn = pass_fn;
+
+    let mut target_widget_id = target;
+    while let Some(widget_id) = target_widget_id {
+        let (widget, pass_ctx) = get_widget_mut(&mut root.widget_arena, widget_id);
+        let parent_id = pass_ctx.parent();
+
+        let mut ctx = LifeCycleCtx {
+            global_state: &mut root.state,
+            widget_state: pass_ctx.widget_state.item,
+            widget_state_children: pass_ctx.widget_state.children,
+            widget_children: pass_ctx.widget_children,
+        };
+        pass_fn(widget, &mut ctx);
+
+        merge_state_up(&mut root.widget_arena, widget_id, root_state);
+        target_widget_id = parent_id;
+    }
+}
+
+pub(crate) fn run_update_pointer_pass(root: &mut RenderRoot, root_state: &mut WidgetState) {
+    let pointer_pos = root.last_mouse_pos.map(|pos| (pos.x, pos.y).into());
+
+    // -- UPDATE HOVERED WIDGETS --
+
+    let hovered_widget = root.state.hovered_path.last().copied();
+    let mut next_hovered_widget = if let Some(pos) = pointer_pos {
+        // TODO - Apply scale?
+        root.get_root_widget()
+            .find_widget_at_pos(pos)
+            .map(|widget| widget.id())
+    } else {
+        None
+    };
+    if let Some(capture_target) = root.state.pointer_capture_target {
+        if next_hovered_widget != Some(capture_target) {
+            next_hovered_widget = None;
+        }
+    }
```

Comment:
```suggestion
    let pointer_capture_target = root.state.pointer_capture_target;
    if pointer_capture_target.is_some() && pointer_capture_target != next_hovered_widget {
        next_hovered_widget = None;
    }
```

(could reuse that `pointer_capture_target` below)

Reply:
In that case it's because I don't want to merge that change. I don't think it improves readability.

It was kind of a dick move to mark it as resolved without explaining the reason, sorry.

I think every other comment you'd made I've either addressed in a follow-up commit or explained why I wouldn't?

Reply:
(Note that sometimes I mark a comment as resolved after I wrote the change that addresses it, but before I've pushed it. It helps me keep track of what to do, but maybe it's not worth the confusion it creates for reviewers.)

Reply:
Yeah, I'm fine with a short comment, explaining that (*although* I disagree), generally I'd prefer, when the author of a review-comment resolves it, unless it's obviously resolved, as (and don't take this personal) it looks like silencing concerns.

> I think every other comment you'd made I've either addressed

No, the nits for example not (although being nits, it's not being really important), but a short answer to e.g. "or is this intentional?" would've been nice.

Reply:
(I also don't think that this suggestion is better)